add translation for web components page (#107)
vuejs-translations/docs-uk: src/guide/extras/web-components.md

# Vue і вебкомпоненти {#vue-and-web-components}

[Вебкомпоненти](https://developer.mozilla.org/en-US/docs/Web/Web_Components) — це загальний термін для набору власних веб-інтерфейсів API, які дозволяють розробникам створювати багаторазові спеціальні елементи.

Ми вважаємо Vue і вебкомпоненти насамперед додатковими технологіями. Vue має відмінну підтримку як для споживання, так і для створення власних елементів. Якщо ви інтегруєте спеціальні елементи в існуючий додаток Vue або використовуєте Vue для створення та розповсюдження спеціальних елементів, ви в хорошій компанії.

## Використання спеціальних елементів у Vue {#using-custom-elements-in-vue}

> Vue [отримав ідеальні 100% у тестах Custom Elements Everywhere](https://custom-elements-everywhere.com/libraries/vue/results/results.html). Використання користувальницьких елементів у програмі Vue здебільшого працює так само, як і використання власних елементів HTML, з деякими речами, про які слід пам’ятати:

Reviewer: Не зовсім 100%, але це питання не до нас)) https://custom-elements-everywhere.com/libraries/vue/results/results.html

> Vue [отримав ідеальні 100% у тестах Custom Elements Everywhere](https://custom-elements-everywhere.com/libraries/vue/results/results.html). Використання користувальницьких елементів у програмі Vue здебільшого працює так само, як і використання власних елементів HTML, з деякими речами, про які слід пам’ятати:

Reviewer: ```suggestion
Vue [отримав ідеальні 100% у тестах Custom Elements Everywhere](https://custom-elements-everywhere.com/libraries/vue/results/results.html). Використання користувацьких елементів у додатку на Vue здебільшого працює так само, як і використання власних елементів HTML, з деякими речами, про які слід пам’ятати:
```

> ### Пропускання роздільної здатності компонента {#skipping-component-resolution}

Reviewer: ```suggestion
### Уникнення розпізнавання компонента {#skipping-component-resolution}
```

Під "resolution" не мається на увазі "роздільна здатність". Мається на увазі розпізнавання його як Vue компонента, яке ми можемо уникнути за допомогою `compilerOptions.isCustomElement`

> За промовчанням Vue намагатиметься розпізнати нерідний HTML-тег як зареєстрований компонент Vue, перш ніж повернутися до відтворення його як спеціального елемента. Це призведе до того, що під час розробки Vue видаватиме попередження «не вдалося вирішити компонент». Щоб повідомити Vue, що певні елементи слід розглядати як спеціальні елементи та пропускати роздільну здатність компонентів, ми можемо вказати параметр [`compilerOptions.isCustomElement`](/api/application.html#app-config-compileroptions).

Reviewer: ```suggestion
За промовчанням Vue намагатиметься розпізнати нерідний HTML-тег як зареєстрований компонент Vue, перш ніж повернутися до відтворення його як спеціального елемента. Це призведе до того, що під час розробки Vue видаватиме попередження «не вдалося визначити компонент» (англ.: "failed to resolve component"). Щоб повідомити Vue, що певні елементи слід розглядати як спеціальні елементи та запобігати їх визначенню як компонентів, ми можемо вказати параметр [`compilerOptions.isCustomElement`](/api/application.html#app-config-compileroptions).
```

> Якщо ви використовуєте Vue із налаштуваннями збірки, цей параметр слід передати через конфігурації збірки, оскільки це параметр під час компіляції.

Reviewer: ```suggestion
Якщо ви використовуєте Vue із налаштуваннями збірки, цей параметр слід передати через конфігурації збірки, оскільки це є параметром часу компіляції.
```

      .tap(options => ({
        ...options,
        compilerOptions: {
          // обробляти будь-які теги, які починаються з ion- як спеціальні елементи
          isCustomElement: tag => tag.startsWith('ion-')
        }
      }))
  }
}
```

### Передача властивостей DOM {#passing-dom-properties}

Оскільки атрибути DOM можуть бути лише рядками, нам потрібно передавати складні дані користувацьким елементам як властивості DOM. Під час встановлення реквізитів для спеціального елемента Vue 3 автоматично перевіряє наявність властивості DOM за допомогою оператора `in` і віддасть перевагу встановленню значення як властивості DOM, якщо ключ присутній. Це означає, що в більшості випадків вам не потрібно думати про це, якщо спеціальний елемент відповідає [рекомендованим найкращим практикам](https://web.dev/custom-elements-best-practices/).

> Однак у рідкісних випадках можуть бути випадки, коли дані потрібно передати як властивість DOM, але настроюваний елемент не визначає/відображає властивість належним чином (що призводить до помилки перевірки `in`). У цьому випадку ви можете примусово встановити прив’язку `v-bind` як властивість DOM за допомогою модифікатора `.prop`:

Reviewer: ```suggestion
Однак, можуть бути рідкісні випадки, коли дані потрібно передати як властивість DOM, але налаштовуваний елемент не визначає/відображає властивість належним чином (що призводить до помилки перевірки `in`). У цьому випадку ви можете примусово встановити прив’язку `v-bind` як властивість DOM за допомогою модифікатора `.prop`:
```

```vue-html
<my-element :user.prop="{ name: 'jack' }"></my-element>

<!-- скорочений еквівалент -->
<my-element .user="{ name: 'jack' }"></my-element>
```

> ## Створення спеціальних елементів за допомогою Vue {#building-custom-elements-with-vue}

Reviewer: ```suggestion
## Створення користувацьких елементів за допомогою Vue {#building-custom-elements-with-vue}
```

import { defineCustomElement } from 'vue'

const MyVueElement = defineCustomElement({
  // звичайні параметри компонента Vue тут
  props: {},
  emits: {},
  template: `...`,

>   // defineCustomElement only: CSS буде введено в тіньовий кореневий елемент

Reviewer: ```suggestion
  // Лише defineCustomElement: CSS буде введено в тіньовий кореневий елемент
```

  styles: [`/* inlined css */`]
})

> // Зареєструйте настроюваний елемент.

Reviewer: ```suggestion
// Зареєструйте користувацький елемент.
```

До речі, помітив, що ти на цій сторінці використовуєш десь "настроюваний", а десь "спеціальний", а ще десь "користувацький".

Author: да, щось наплутав

// Після реєстрації всі теги `<my-vue-element>`
// на сторінці буде оновлено.
customElements.define('my-vue-element', MyVueElement)

// Ви також можете програмно створити екземпляр елемента:
// (можливо лише після реєстрації)
document.body.appendChild(
  new MyVueElement({
    // початкові реквізити (необов'язково)
  })
)
```

#### Життєвий цикл {#Lifecycle}

> - Спеціальний елемент Vue монтуватиме внутрішній екземпляр компонента Vue у своєму тіньовому корені, коли елемент [`connectedCallback`](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#using_the_lifecycle_callbacks) дзвонив вперше.

Reviewer: ```suggestion
- Спеціальний елемент Vue змонтує внутрішній екземпляр компонента Vue у своєму тіньовому корені, коли елемент [`connectedCallback`](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#using_the_lifecycle_callbacks) викликаний вперше.
```

- Коли викликається `disconnectedCallback` елемента, Vue перевірить, чи елемент від’єднано від документа після позначки мікрозавдання.

>    - Якщо елемент все ще знаходиться в документі, це переміщення, і екземпляр компонента буде збережено;

Reviewer: ```suggestion
   - Якщо елемент все ще знаходиться в документі — це переміщення, тому екземпляр компонента буде збережено;
```

>    - Якщо елемент від’єднано від документа, це буде видалення, і екземпляр компонента буде демонтовано.

Reviewer: ```suggestion
   - Якщо елемент від’єднано від документа — це видалення, тому екземпляр компонента буде демонтовано.
```

#### Реквізити {#Props}

- Усі реквізити, оголошені за допомогою параметра `props`, будуть визначені в користувацькому елементі як властивості. Vue автоматично оброблятиме відображення між атрибутами / властивостями, де це необхідно.

   - Атрибути завжди відображаються у відповідних властивостях.

>    - Реквізити з примітивними значеннями (`рядок`, `логічний` або `число`) відображаються як атрибути.

Reviewer: ```suggestion
   - Реквізити з примітивними значеннями (`string`, `boolean` або `number`)відображаються як атрибути.
```
Конструкції мови JS не перекладаємо

  }
  ```

>   І використання спеціального елемента:

Reviewer: ```suggestion
  І використання користувацького елемента:
```

  ```vue-html
  <my-element selected index="1"></my-element>
  ```

  У компоненті `selected` буде приведено до `true` (логічне значення), а `index` — до `1` (число).

#### Події {#Events}

> Події, створені через `this.$emit` або налаштування `emit`, надсилаються як власні [CustomEvents](https://developer.mozilla.org/en-US/docs/Web/EventsCreating_and_triggering_events#adding_custom_data_%E2%80%93_customevent) на спеціальному елементі. Додаткові аргументи події (корисне навантаження) відображатимуться як масив об’єкта CustomEvent як його властивість `detail`.

Reviewer: ```suggestion
Події, створені через `this.$emit` або `emit` в `setup`, надсилаються як рідні [CustomEvents](https://developer.mozilla.org/en-US/docs/Web/EventsCreating_and_triggering_events#adding_custom_data_%E2%80%93_customevent) на користувацькому елементі. Додаткові аргументи події (корисне навантаження) відображатимуться як масив об’єкта CustomEvent як його властивість `detail`.
```

#### Слоти {#Slots}

> Усередині компонента слоти можна відобразити за допомогою елемента `<slot/>`, як зазвичай. Однак під час використання отриманого елемента він приймає лише [власний синтаксис слотів](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots):

Reviewer: ```suggestion
Усередині компонента слоти можна відобразити за допомогою елемента `<slot/>`, як зазвичай. Однак під час використання отриманого елемента він приймає лише [рідний синтаксис слотів](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots):
```
Для консистентності — десь вище "native" ти перекладав як "рідний"

- [Обмежені слоти](/guide/components/slots.html#scoped-slots) не підтримуються.

- При передачі іменованих слотів використовуйте атрибут `slot` замість директиви `v-slot`:

  ```vue-html
  <my-element>
    <div slot="named">привіт</div>
  </my-element>
  ```

#### Надати / ввести {#provideinject}

> [Надати / ввести API](/guide/components/provide-inject.html#provide-inject) і [еквівалент композиційного АРІ](/api/composition-api-dependency-injection.html#provide) також працюють між Vue - визначені спеціальні елементи. Однак зауважте, що це працює **тільки між настроюваними елементами**. тобто визначений спеціальний елемент Vue не зможе вводити властивості, надані компонентом Vue, що не є настроюваним елементом.

Reviewer: ```suggestion
[API надавання / введення](/guide/components/provide-inject.html#provide-inject) і [еквівалент композиційного АРІ](/api/composition-api-dependency-injection.html#provide) також працюють з користувацькими елементи, визначеними Vue. Однак зауважте, що це працює **тільки з користувацькими елементами**. Тобто користувацький елемент, визначений Vue, не зможе вводити властивості, надані компонентом Vue, що не є користувацьким елементом.
```

### SFC як спеціальний елемент {#sfc-as-custom-element}

> `defineCustomElement` також працює з однофайловими компонентами Vue (SFC). Однак із налаштуванням інструментів за промовчанням `<style>` всередині SFC все одно буде витягнуто та об’єднано в один файл CSS під час робочої збірки. У разі використання SFC як користувацького елемента, часто бажано замість цього додати теги `<style>` до тіньового кореня користувацького елемента.

Reviewer: ```suggestion
`defineCustomElement` також працює з однофайловими компонентами Vue (SFC). Однак із налаштуванням інструментів за промовчанням `<style>` всередині SFC все одно буде витягнуто та об’єднано в один файл CSS під час виробничої збірки. У разі використання SFC як користувацького елемента, часто бажано замість цього додати теги `<style>` до тіньового кореня користувацького елемента.
```
Ми десь вже "production" перекладаємо як "виробництво". Додав в глосарій, щоб було однаково

> Офіційні інструменти SFC підтримують імпорт SFC у «режимі спеціальних елементів» (потрібно `@vitejs/plugin-vue@^1.4.0` або `vue-loader@^16.5.0`). SFC, завантажений у режимі спеціального елемента, вставляє свої теги `<style>` як рядки CSS і розкриває їх у параметрі `styles` компонента. Це буде підібрано `defineCustomElement` і вставлено в тіньовий корінь елемента під час створення екземпляра.

Reviewer: ```suggestion
Офіційні інструменти SFC підтримують імпорт SFC у «режимі користувацьких елементів» (потрібно `@vitejs/plugin-vue@^1.4.0` або `vue-loader@^16.5.0`). SFC, завантажений у режимі користувацького елемента, вставляє свої теги `<style>` як рядки CSS і розкриває їх у параметрі `styles` компонента. Це буде підібрано `defineCustomElement` і вставлено в тіньовий корінь елемента під час створення екземпляра.
```

Щоб увімкнути цей режим, просто завершіть назву файлу компонента `.ce.vue`:

```js
import { defineCustomElement } from 'vue'
import Example from './Example.ce.vue'

console.log(Example.styles) // ["/* вбудований css */"]

> // перетворити на власний конструктор елементів

Reviewer: ```suggestion
// перетворити на користувацький конструктор елементів
```

const ExampleElement = defineCustomElement(Example)

> // зареєструвати

Reviewer: ```suggestion
// зареєструйте
```
(так само було перекладено вище)

customElements.define('my-example', ExampleElement)
```

Якщо ви бажаєте налаштувати, які файли слід імпортувати в режимі спеціальних елементів (наприклад, розглядаючи _all_ SFC як спеціальні елементи), ви можете передати опцію `customElement` до відповідних плагінів збірки:

- [@vitejs/plugin-vue](https://github.com/vitejs/vite/tree/main/packages/plugin-vue#using-vue-sfcs-as-custom-elements)
- [vue-loader](https://github.com/vuejs/vue-loader/tree/next#v16-only-options)

> ### Поради щодо бібліотеки спеціальних елементів Vue {#tips-for-a-vue-custom-elements-library}

Reviewer: ```suggestion
### Поради щодо бібліотеки користувацьких елементів Vue {#tips-for-a-vue-custom-elements-library}
```

> Під час створення спеціальних елементів за допомогою Vue, елементи покладатимуться на середовище виконання Vue. Існує вартість базового розміру ~16 Кб залежно від того, скільки функцій використовується. Це означає, що Vue не ідеально використовувати, якщо ви надсилаєте окремий спеціальний елемент – ви можете використовувати ванільний JavaScript, [petite-vue](https://github.com/vuejs/petite-vue) або фреймворки, які спеціалізуються на невеликому розмірі часу виконання. Однак базовий розмір більш ніж виправданий, якщо ви постачаєте колекцію спеціальних елементів зі складною логікою, оскільки Vue дозволить створювати кожен компонент із набагато меншим кодом. Чим більше елементів ви доставляєте разом, тим кращий компроміс.

Reviewer: ```suggestion
Під час створення користувацьких елементів за допомогою Vue, елементи покладатимуться на середовище виконання Vue. Існує вартість базового розміру ~16 Кб залежно від того, скільки функцій використовується. Це означає, що Vue не ідеально використовувати, якщо ви надсилаєте окремий користувацький елемент – ви можете використовувати ванільний JavaScript, [petite-vue](https://github.com/vuejs/petite-vue) або фреймворки, які спеціалізуються на невеликому розмірі часу виконання. Однак базовий розмір більш ніж виправданий, якщо ви постачаєте колекцію користувацьких елементів зі складною логікою, оскільки Vue дозволить створювати кожен компонент із набагато меншим кодом. Чим більше елементів ви доставляєте разом, тим кращий компроміс.
```

> Якщо спеціальні елементи використовуватимуться в програмі, яка також використовує Vue, ви можете вибрати екстерналізацію Vue із вбудованого набору, щоб елементи використовували ту саму копію Vue із головної програми.

Reviewer: ```suggestion
Якщо користувацькі елементи застосовуються в програмі, яка також використовує Vue, ви можете вибрати екстерналізацію Vue із вбудованого набору, щоб елементи використовували ту саму копію Vue із головної програми.
```

Модель компонентів Vue розроблена з урахуванням цих потреб як узгоджена система.

Маючи компетентну команду інженерів, ви, ймовірно, зможете створити еквівалент на основі власних спеціальних елементів, але це також означає, що ви берете на себе довготерміновий тягар обслуговування внутрішньої структури, водночас втрачаючи переваги екосистеми та спільноти зрілий фреймворк, такий як Vue.

Існують також фреймворки, побудовані з використанням спеціальних елементів як основи їх компонентної моделі, але всі вони неминуче мають запроваджувати власні рішення проблем, перелічених вище. Використання цих фреймворків передбачає прийняття їхніх технічних рішень щодо вирішення цих проблем — що, незважаючи на те, що може бути рекламоване, автоматично не захищає вас від потенційних майбутніх відтоків.

Є також деякі області, де ми вважаємо, що спеціальні елементи є обмеженнями:

> - Ретельне оцінювання слотів перешкоджає створенню компонентів. [Обмежені слоти](/guide/components/slots.html#scoped-slots) Vue — це потужний механізм композиції компонентів, який не підтримується спеціальними елементами через потребу нативних слотів. Охочі слоти також означають, що приймаючий компонент не може контролювати, коли та чи відтворювати частину вмісту слота.

Reviewer: ```suggestion
- Ретельне оцінювання слотів перешкоджає створенню компонентів. [Обмежені слоти](/guide/components/slots.html#scoped-slots) Vue — це потужний механізм композиції компонентів, який не підтримується користувацькими елементами через потребу рідних слотів. Охочі слоти також означають, що приймаючий компонент не може контролювати, коли та чи відтворювати частину вмісту слота.
```

> - Для доставки спеціальних елементів із CSS із тіньовою областю DOM сьогодні потрібно вбудовувати CSS у JavaScript, щоб їх можна було вставляти в тіньові корені під час виконання. Вони також призводять до дублювання стилів у розмітці в сценаріях SSR. У цій сфері працюють над [функціями платформи](https://github.com/whatwg/html/pull/4898/), але наразі вони ще не підтримуються повсюдно, і все ще є робоча продуктивність / SSR проблеми, які необхідно вирішити. Тим часом SFC Vue надають [механізми визначення області CSS](/api/sfc-css-features.html), які підтримують вилучення стилів у звичайні файли CSS.

Reviewer: ```suggestion
- Для доставки користувацьких елементів із CSS із тіньовою областю DOM сьогодні потрібно вбудовувати CSS у JavaScript, щоб їх можна було вставляти в тіньові корені під час виконання. Вони також призводять до дублювання стилів у розмітці в сценаріях SSR. У цій сфері працюють над [функціями платформи](https://github.com/whatwg/html/pull/4898/), але наразі вони ще не підтримуються всюди, і все ще є проблеми з робочою продуктивністю / SSR, які необхідно вирішити. Тим часом Vue SFC надає [механізми визначення області CSS](/api/sfc-css-features.html), які підтримують вилучення стилів у звичайні файли CSS.
```
